packages/beacon-node/src/chain/rewards/blockRewards.ts

```diff
@@ -0,0 +1,130 @@
+  total: SubRewardValue;
+  attestations: SubRewardValue;
+  syncAggregate: SubRewardValue;
+  proposerSlashings: SubRewardValue;
+  attesterSlashings: SubRewardValue;
+};
+
+/**
+ * Calculate total proposer block rewards given block and the beacon state of the same slot before the block is applied
+ * Standard (Non MEV) rewards for proposing a block consists of:
+ *  1) Including attestations from (beacon) committee
+ *  2) Including attestations from sync committee
+ *  3) Reporting slashable behaviours from proposer and attester
+ */
+export async function computeBlockRewards(
+  block: allForks.BeaconBlock,
+  state: CachedBeaconStateAllForks
+): Promise<BlockRewards> {
+  const fork = state.config.getForkName(block.slot);
+  const blockAttestationReward =
+    fork === ForkName.phase0
+      ? computeBlockAttestationRewardPhase0(block as phase0.BeaconBlock, state as CachedBeaconStatePhase0)
+      : computeBlockAttestationRewardAltair(block as altair.BeaconBlock, state as CachedBeaconStateAltair);
+  const syncAggregateReward = computeSyncAggregateReward(block as altair.BeaconBlock, state as CachedBeaconStateAltair);
+  const blockProposerSlashingReward = computeBlockProposerSlashingReward(block, state);
+  const blockAttesterSlashingReward = computeBlockAttesterSlashingReward(block, state);
+
+  const total =
+    blockAttestationReward + syncAggregateReward + blockProposerSlashingReward + blockAttesterSlashingReward;
+
+  return {
+    proposerIndex: block.proposerIndex,
+    total,
+    attestations: blockAttestationReward,
+    syncAggregate: syncAggregateReward,
+    proposerSlashings: blockProposerSlashingReward,
+    attesterSlashings: blockAttesterSlashingReward,
+  };
+}
+
+/**
+ * TODO: Calculate rewards received by block proposer for incuding attestations.
+ */
+function computeBlockAttestationRewardPhase0(
+  _block: phase0.BeaconBlock,
+  _state: CachedBeaconStatePhase0
+): SubRewardValue {
+  throw new Error("Unsupported fork! Block attestation reward calculation is not yet available in phase0");
+}
+
+/**
+ * Calculate rewards received by block proposer for incuding attestations since Altair.
+ * Reuses `processAttestationsAltair()`. Has dependency on RewardCache
+ */
+function computeBlockAttestationRewardAltair(
+  block: altair.BeaconBlock,
+  state: CachedBeaconStateAltair
+): SubRewardValue {
+  const fork = state.config.getForkSeq(block.slot);
+  const {attestations} = block.body;
+
+  processAttestationsAltair(fork, state, attestations, false);
+
+  return state.proposerRewards.attestations;
+}
+
+function computeSyncAggregateReward(block: altair.BeaconBlock, state: CachedBeaconStateAltair): SubRewardValue {
+  if (block.body.syncAggregate !== undefined) {
+    const {syncCommitteeBits} = block.body.syncAggregate;
+    const {syncProposerReward} = state.epochCtx;
+
+    return syncCommitteeBits.getTrueBitIndexes().length * Math.floor(syncProposerReward); // syncProposerReward should already be integer
```

Comment:
this only count proposer reward for participant not penalties for non-participant. I think it's missing in this too https://github.com/ChainSafe/lodestar/blob/61cf1a87600a1b756fd46d314821f6987d3d8b42/packages/state-transition/src/block/processSyncCommittee.ts#L48

@ensi321 can you correct the proposer reward in `processSyncAggregate` and consume it here to avoid duplicate the logic?

Reply:
since we have the `proposerRewards` cache, why don't we call `processBlock()` and get data from the cache? if the block state is available, we even can get `proposerRewards` from there without having to `processBlock`.

Reply:
> this only count proposer reward for participant not penalties for non-participant. I think it's missing in this too
> 
> https://github.com/ChainSafe/lodestar/blob/61cf1a87600a1b756fd46d314821f6987d3d8b42/packages/state-transition/src/block/processSyncCommittee.ts#L48
> 
> @ensi321 can you correct the proposer reward in `processSyncAggregate` and consume it here to avoid duplicate the logic?

The sync committee penalty is only applied to those non-participants. The reward for proposer is simply number of sync committee participant * weight. 

Here are what [teku](https://github.com/Consensys/teku/blob/5c2c4a841b6e8ec510189e48b361581f0107bc0e/data/provider/src/main/java/tech/pegasys/teku/api/RewardCalculator.java#L187) and [lighthouse](https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/beacon_node/beacon_chain/src/block_reward.rs#L108) implementation for reference.

Reply:
> since we have the `proposerRewards` cache, why don't we call `processBlock()` and get data from the cache? if the block state is available, we even can get `proposerRewards` from there without having to `processBlock`.

I think `computeBlockRewards()` can first check on `proposerRewards` cache. If not, we need to explicitly calculate the rewards. 

I think some of the rewards (eg. sync reward) is simple enough to calculate in one line of code. It is cleaner and more explicit to state the reward formula here then going through the entire `processSyncAggregate()`.